go/oasis-node/cmd/debug/dumpdb/dumpdb.go

```diff
@@ -0,0 +1,383 @@
+	storageDB "github.com/oasislabs/oasis-core/go/storage/database"
+
+	"encoding/json"
+)
+
+const (
+	cfgDumpOutput     = "dump.output"
+	cfgDumpReadOnlyDB = "dump.read_only_db"
+	cfgDumpVersion    = "dump.version"
+)
+
+var (
+	dumpDBCmd = &cobra.Command{
+		Use:   "dumpdb",
+		Short: "dump the on-disk consensus DB to a JSON document",
+		Run:   doDumpDB,
+	}
+
+	dumpDBFlags = flag.NewFlagSet("", flag.ContinueOnError)
+
+	logger = logging.GetLogger("cmd/debug/dumpdb")
+)
+
+func doDumpDB(cmd *cobra.Command, args []string) {
+	var ok bool
+	defer func() {
+		if !ok {
+			os.Exit(1)
+		}
+	}()
+
+	if err := cmdCommon.Init(); err != nil {
+		cmdCommon.EarlyLogAndExit(err)
+	}
+
+	dataDir := cmdCommon.DataDir()
+	if dataDir == "" {
+		logger.Error("data directory must be set")
+		return
+	}
+
+	// Load the old genesis document, required for filling in parameters
+	// that are not persisted to ABCI state.
+	fp, err := genesisFile.NewFileProvider(flags.GenesisFile())
+	if err != nil {
+		logger.Error("failed to load existing genesis document",
+			"err", err,
+		)
+		return
+	}
+	oldDoc, err := fp.GetGenesisDocument()
+	if err != nil {
+		logger.Error("failed to get existing genesis document",
+			"err", err,
+		)
+		return
+	}
+
+	// Initialize the ABCI state storage for access.
+	//
+	// Note: While it would be great to always use read-only DB access,
+	// badger will refuse to open a DB that isn't closed properly in
+	// read-only mode because it needs to truncate the value log.
+	//
+	// Hope you have backups if you ever run into this.
+	ctx := context.Background()
+	ldb, _, stateRoot, err := abci.InitStateStorage(
+		ctx,
+		&abci.ApplicationConfig{
+			DataDir:           filepath.Join(dataDir, tendermint.StateDir),
+			StorageBackend:    storageDB.BackendNameBadgerDB, // No other backend for now.
+			MemoryOnlyStorage: false,
+			ReadOnlyStorage:   viper.GetBool(cfgDumpReadOnlyDB),
+		},
+	)
+	if err != nil {
+		logger.Error("failed to initialize ABCI storage backend",
+			"err", err,
+		)
+	}
+	defer ldb.Cleanup()
+
+	latestVersion := int64(stateRoot.Version)
+	dumpVersion := viper.GetInt64(cfgDumpVersion)
+	if dumpVersion == 0 {
+		dumpVersion = latestVersion
+	}
+	if dumpVersion <= 0 || dumpVersion > latestVersion {
+		logger.Error("dump requested for version that does not exist",
+			"dump_version", dumpVersion,
+			"latest_version", latestVersion,
+		)
+		return
+	}
+
+	// Generate the dump by querying all of the relevant backends, and
+	// extracting the immutable parameters from the current genesis
+	// document.
+	//
+	// WARNING: The state is not guaranteed to be usable as a genesis
+	// document without manual intervention, and only the state that
+	// would be exported by the normal dump process will be present
+	// in the dump.
+	qs := &dumpQueryState{
+		ldb:    ldb,
+		height: dumpVersion,
+	}
+	doc := &genesis.Document{
+		Height:    qs.BlockHeight(),
+		Time:      time.Now(), // XXX: Make this deterministic?
```

Comment:
Making this deterministic would require looking into the Tendermint block store and fetching the corresponding block header. I had some code to do this stashed somewhere, I'll try to dig it up.

Reply:
To be honest since any documents generated by the tool will require manual intervention anyway, I'm ok with this just being the dump time for now.

Reply:
That's fine.